source/ConditionsStore.cpp

```diff
@@ -0,0 +1,276 @@
+bool ConditionsStore::AddCondition(const string &name, int64_t value)
+{
+	// This code performes 2 lookups of the condition, once for get and
+	// once for set. This might be optimized to a single lookup in a
+	// later version of the code when we add derived conditions.
+	
+	return SetCondition(name, GetCondition(name) + value);
+}
+
+
+
+// Set a value for a condition, either for the local value, or by performing
+// a set on the provider.
+bool ConditionsStore::SetCondition(const string &name, int64_t value)
+{
+	ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+	{
+		(storage[name]).value = value;
+		return true;
+	}
+	
+	if(ce->type == VALUE)
+	{
+		ce->value = value;
+		return true;
+	}
+	
+	return (ce->provider).setFun(name, value);
+}
+
+
+
+// Erase a condition completely, either the local value or by performing
+// an erase on the provider.
+bool ConditionsStore::EraseCondition(const string &name)
+{
+	ConditionEntry *ce = GetEntry(name);
+	if(!ce)
+		return true;
+	
+	if(ce->type == VALUE)
+	{
+		storage.erase(name);
+		return true;
+	}
+	
+	return (ce->provider).eraseFun(name);
+}
+
+
+
+ConditionsStore::PrimariesIterator ConditionsStore::PrimariesBegin() const
+{
+	return PrimariesIterator(storage.begin(), storage.end());
+}
+
+
+
+ConditionsStore::PrimariesIterator ConditionsStore::PrimariesEnd() const
+{
+	return PrimariesIterator(storage.end(), storage.end());
+}
+
+
+
+ConditionsStore::PrimariesIterator ConditionsStore::PrimariesLowerBound(const string &key) const
+{
+	return PrimariesIterator(storage.lower_bound(key), storage.end());
+}
+
+
+
+// Sets a provider for a given prefix.
+void ConditionsStore::SetProviderPrefixed(const string &prefix, DerivedProvider conditionsProvider)
+{
+	storage[prefix].provider = conditionsProvider;
+	storage[prefix].type = PREFIX_PROVIDER;
+}
+
+
+
+// Sets a provider for the condition identified by the given name.
+void ConditionsStore::SetProviderNamed(const string &name, DerivedProvider conditionProvider)
+{
+	storage[name].provider = conditionProvider;
+	storage[name].type = EXACT_PROVIDER;
+}
+
+
+
+// Helper to completely remove all data and linked condition-providers from the store.
+void ConditionsStore::Clear()
+{
+	storage.clear();
+}
+
+
+
+ConditionsStore::ConditionEntry *ConditionsStore::GetEntry(const string &name)
+{
+	if(storage.empty())
+		return nullptr;
+	
+	// Perform a single search for values, named providers and prefixed providers.
+	auto it = storage.upper_bound(name);
+	if(it == storage.begin())
+		return nullptr;
+	
+	--it;
+	// The entry is valid if we have an exact stringmatch, but also when we have a
+	// prefix entry and the prefix part matches.
+	if(!(name.compare(0, it->first.length(), it->first)) &&
+			(it->second.type == PREFIX_PROVIDER || it->first.length() == name.length()))
+		return &(it->second);
+	
+	return nullptr;
+}
+
+
+
+const ConditionsStore::ConditionEntry *ConditionsStore::GetEntryConst(const string &name) const
```

Comment:
Why not just `GetEntry`? Why the extra `Const` in the name?

Reply:
Agreed, the extra Const in the name is probably not needed, I'll check if I can remove it.

Reply:
Also, since the 2 functions are the same I'd consider using the `const_cast` hack if it doesn't violate the style guide :)